### PR TITLE
👷(circle) move to a richie's container based CI

### PR DESCRIPTION
**Pull request by jmaupetit**
- [x] Add docker-compose CI configuration
- [x] Build the production & dev container
- [x] Use the dev container to run back-end tests
- [x] Test the front-end with a node-based container (not the dev container?)
- [x] Build & test alpine based images
- [x] Push all container flavors to docker hub upon a new release (tag)
- [x] Document commit message
- [x] Update CI required job names in repository settings

### CODE REVIEW

**Comment by sampaccoud** [comment]
I think the commit message deserves a little more love :-)

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,84 +1,375 @@
...
-      - restore_cache:
-          keys:
-          - v3-back-dependencies-{{ checksum "requirements/base.txt" }}-{{ checksum "requirements/dev.txt" }}
-          # fallback to using the latest cache if no exact match is found
-          - v3-back-dependencies-
+
+      # Each image is tagged with the current git commit sha1 to avoid
+      # collisions in parallel builds.
+      - run:
+          name: Build production container
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .circleci/config.yml]: Totally :+1:

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,84 +1,375 @@
...
+          command: |
+            docker build \
+              -t richie:${CIRCLE_SHA1} \
+              .
+
+      - run:
+          name: List available images
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .circleci/config.yml]: What about:

> Check built image availability

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,84 +1,375 @@
...
+            - ~/fun/docker/images/
+          key: docker-images-v1-{{ .Revision }}-dev
 
   lint-back-flake8:
-    docker:
-      - image: python:3.6-stretch
-    working_directory: ~/richie
+
+    machine:
+      docker_layer_caching: true
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .circleci/config.yml]: Yup totally. Thx!

File: .circleci/config.yml [hunk 4]
```diff
@@ -134,41 +424,84 @@ workflows:
...
+      - hub:
+          requires:
+            - test-front
+            - test-back
+            - test-alpine
+          filters:
+            branches:
+              ignore: /.*/
+            tags:
+              only: /^v.*/
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .circleci/config.yml]: Do you think we will need a more complicated pattern than semantic versioning for Richie (with the `v` prefix for filtering)?

**Comment by jmaupetit** [comment]
> I think the commit message deserves a little more love :-)

Yup, it's the latest mark to check on the road map.

**Pull request merged**